File: r/tests/test_factor_variables.R
#!/usr/bin/env Rscript
# Test factor variable expansion

library(polars)
library(duckdb)
library(DBI)
source("R/common.R")
source("R/polars.R")
source("R/duckdb.R")

set.seed(42)
n <- 1000

cat("Testing factor variable expansion...\n\n")

# Test 1: Factor variables with Polars
cat("Test 1: Factor variables (Polars)... ")
df <- pl$DataFrame(
  y = rnorm(n),
  x = rnorm(n),
  region = sample(c("A", "B", "C"), n, replace = TRUE),
  fe1 = sample(0:9, n, replace = TRUE),
  fe2 = sample(0:4, n, replace = TRUE)
)

result <- leanfe_polars(df, formula = "y ~ x + i(region) | fe1 + fe2")

# Should have coefficient for x plus 2 region dummies (A is reference)
stopifnot("x" %in% names(result$coefficients))
stopifnot("region_B" %in% names(result$coefficients))
stopifnot("region_C" %in% names(result$coefficients))
stopifnot(length(result$coefficients) == 3)  # x + 2 region dummies
cat("PASSED\n")
cat(sprintf("  Coefficients: %s\n", paste(names(result$coefficients), collapse = ", ")))
cat("  (region_A is reference category)\n")

# Test 2: Factor variables with DuckDB
cat("Test 2: Factor variables (DuckDB)... ")
result <- leanfe_duckdb(df, formula = "y ~ x + i(region) | fe1 + fe2")

stopifnot("x" %in% names(result$coefficients))
stopifnot("region_B" %in% names(result$coefficients))
stopifnot("region_C" %in% names(result$coefficients))
stopifnot(length(result$coefficients) == 3)
cat("PASSED\n")
cat(sprintf("  Coefficients: %s\n", paste(names(result$coefficients), collapse = ", ")))
cat("  (region_A is reference category)\n")

cat("\n✅ All 2 tests passed!\n")

File: python/tests/test_factor_variables.py
import polars as pl
import numpy as np
from leanfe import leanfe_polars, leanfe_duckdb

def test_factor_variables_polars():
    """Test factor variable expansion into dummies (first category dropped as reference)"""
    np.random.seed(42)
    n = 1000
    
    df = pl.DataFrame({
        'y': np.random.randn(n),
        'x': np.random.randn(n),
        'region': np.random.choice(['A', 'B', 'C'], n),
        'fe1': np.random.randint(0, 10, n),
        'fe2': np.random.randint(0, 5, n)
    })
    
    # Test with factor variable (expanded to dummies, first category dropped)
    result = leanfe_polars(df, formula="y ~ x + i(region) | fe1 + fe2")
    
    # Should have coefficient for x plus 2 region dummies (A is reference)
    assert 'x' in result['coefficients']
    assert 'region_B' in result['coefficients']
    assert 'region_C' in result['coefficients']
    assert len(result['coefficients']) == 3  # x + 2 region dummies
    
    print("✓ Polars factor variables test passed")
    print(f"  Coefficients: {list(result['coefficients'].keys())}")
    print(f"  (region_A is reference category)")

def test_factor_variables_duckdb():
    """Test factor variable expansion into dummies with DuckDB"""
    np.random.seed(42)
    n = 1000
    
    df = pl.DataFrame({
        'y': np.random.randn(n),
        'x': np.random.randn(n),
        'region': np.random.choice(['A', 'B', 'C'], n),
        'fe1': np.random.randint(0, 10, n),
        'fe2': np.random.randint(0, 5, n)
    })
    
    # Test with factor variable (expanded to dummies, first category dropped)
    result = leanfe_duckdb(df, formula="y ~ x + i(region) | fe1 + fe2")
    
    # Should have coefficient for x plus 2 region dummies (A is reference)
    assert 'x' in result['coefficients']
    assert 'region_B' in result['coefficients']
    assert 'region_C' in result['coefficients']
    assert len(result['coefficients']) == 3
    
    print("✓ DuckDB factor variables test passed")
    print(f"  Coefficients: {list(result['coefficients'].keys())}")
    print(f"  (region_A is reference category)")

if __name__ == "__main__":
    test_factor_variables_polars()
    test_factor_variables_duckdb()
    print("\n✓ All factor variable tests passed!")

